### PR TITLE
feat(typing): add type hints to compiled router

### PR DESCRIPTION
**Pull request by CaselIT**
This is mainly to try how cython works with pep484 types

NOTE: this is currently unmergiable, since it's not py 3.5 compatible

Gist with the compilations with cython https://gist.github.com/CaselIT/f0493d74422cce46c95c10e8abd1e4f0

### CODE REVIEW

**Comment by vytas7** [comment]
This should be good to revisit again once https://github.com/falconry/falcon/pull/1947 gets merged.

File: falcon/routing/compiled.py [hunk 1]
```diff
@@ -27,8 +28,12 @@
 from falcon.util.sync import _should_wrap_non_coroutines
 from falcon.util.sync import wrap_sync_to_async
 
-if False:  # TODO: switch to TYPE_CHECKING once support for py3.5 is dropped
-    from typing import Any
+if TYPE_CHECKING:  # TODO: switch to TYPE_CHECKING once support for py3.5 is dropped
```
[the remark this answers is not in the document]
Comment (vytas7) [reply falcon/routing/compiled.py:31]: No longer relevant now that we've dropped < 3.7.

**Comment by vytas7** [comment]
@CaselIT Now we can continue with this one; we just need to address merge conflicts and it may possible to move this forward off _Draft_.

**Comment by CaselIT** [comment]
I'll try getting into it soonish

**Review by vytas7 — APPROVED** [review]
Looks good, and it seems it passes Mypy and all the tests now :100:

**Pull request merged**